File: C++/Problems/Problems/problem_1733.cpp
#include "problem_1733.h"

#include <algorithm>

bool problem_1733::test() {
    int n = 2;
    std::vector<std::vector<int>> languages { { 1 }, { 2 }, { 1, 2 } };
    std::vector<std::vector<int>> friendships { { 1, 2 }, { 1, 3 }, { 2, 3 } };

    int expected = 1;

    int result = minimumTeachings(n, languages, friendships);

    std::cout << result << std::endl;

    return expected == result;
}

int problem_1733::minimumTeachings(int n, std::vector<std::vector<int>> &languages,
    std::vector<std::vector<int>> &friendships) {
    int numPeople = languages.size();
    std::unordered_set<int> nonCommon;
    std::vector<std::unordered_set<int>> languagesSets(numPeople, std::unordered_set<int>());
    for (int i = 0; i < numPeople; ++i)
    {
        languagesSets[i] = std::unordered_set<int>(languages[i].begin(), languages[i].end());
    }
    for (auto friendship : friendships)
    {
        if (isCommonLanguage(languagesSets[friendship[0] - 1], languagesSets[friendship[1] - 1]))
            continue;
        nonCommon.insert(friendship[0] - 1);
        nonCommon.insert(friendship[1] - 1);
    }

    std::vector counts(n, 0);
    for (int non : nonCommon)
    {
        for (int language : languages[non])
        {
            counts[language - 1]++;
        }
    }

    return nonCommon.size() - *std::ranges::max_element(counts);
}


bool problem_1733::isCommonLanguage(std::unordered_set<int> languagesFirst, std::unordered_set<int> languagesSecond) {
    for (int lang : languagesFirst)
    {
        if (languagesSecond.contains(lang))
            return true;
    }

    return false;
}

File: Python/Problems/Problems/problem1733.py
from typing import List, Set

from .problem import Problem


class Problem1733(Problem):
    def test(self) -> bool:
        n = 2
        languages = [[1], [2], [1, 2]]
        friendships = [[1, 2], [1, 3], [2, 3]]

        expected = 1

        result = self.minimumTeachings(n, languages, friendships)

        print(result)

        return expected == result

    def minimumTeachings(self, n: int, languages: List[List[int]], friendships: List[List[int]]) -> int:
        numPeople = len(languages)
        nonCommon = set()
        languagesSets = [set(languages[i]) for i in range(numPeople)]
        for friendship in friendships:
            if self.isCommonLanguage(languagesSets[friendship[0] - 1], languagesSets[friendship[1] - 1]):
                continue
            nonCommon.add(friendship[0] - 1)
            nonCommon.add(friendship[1] - 1)

        counts = [0] * n
        for non in nonCommon:
            for language in languages[non]:
                counts[language - 1] += 1

        return len(nonCommon) - max(counts)

    def isCommonLanguage(self, languagesFirst: Set[int], languagesSecond: Set[int]) -> bool:
        for lang in languagesFirst:
            if lang in languagesSecond:
                return True

        return False